### PR TITLE
Refactor implicit flow.

### PR DESCRIPTION
**Pull request by jimabramson**
Access token handling methods have been moved to
a shared mixin class so they can be reused in both
Authorization and Access Token endpoints.  This
facilitates overriding common behaviors in client
implementations (e.g. OIDC)

ECOM-2440

@clintonb @nasthagiri please review
@justinabrahms @rlucioni FYI

### CODE REVIEW

**Comment by jimabramson** [comment]
I'll take steps to placate coverage failures once I get some feedback on the refactor itself.

File: provider/views.py [hunk 1]
```diff
@@ -255,7 +327,36 @@ def error_response(self, request, error, **kwargs):
...
+
+        if constants.SINGLE_ACCESS_TOKEN:
+            token = self.get_access_token(request, **lookup_kwargs)
+        else:
+            token = self.create_access_token(request, **lookup_kwargs)
+
+        response_data = self.access_token_response_data(token, data['response_type'])
+
+        basepath = data.get("redirect_uri")
+        if not basepath:
```
Comment (justinabrahms) [inline provider/views.py:349]: It is very important that the user cannot redirect to a URL that is not the underlying redirect_uri as defined in the model. If the redirect_uri differs (by how much, I'm unsure. I'd need to re-read the spec), we must return an error.

File: provider/oauth2/views.py [hunk 2]
```diff
@@ -13,10 +13,53 @@
...
+            rt.delete()
+        else:
+            rt.expired = True
+            rt.save()
+
+    def invalidate_access_token(self, at):
+        if constants.DELETE_EXPIRED:
+            at.delete()
+        else:
+            at.expires = now() - timedelta(days=1)
```
Comment (clintonb) [inline provider/oauth2/views.py]: Minor: subtract 1 millisecond.

File: provider/views.py (hunk 1 above)
Comment (jimabramson) [inline provider/views.py:349]: @justinabrahms thanks for calling that out.  I hadn't (intentionally) modified the validation, which I believe was already handled [here](https://github.com/edx/django-oauth2-provider/blob/jsa/implicit2/provider/oauth2/forms.py#L162).  Will double check for loopholes.

**Comment by clintonb** [comment]
@jimabramson so far, so good. I have no objections.

File: provider/views.py [hunk 3]
```diff
@@ -511,46 +551,12 @@ def error_response(self, error, status=400, **kwargs):
...
         response_data = self.access_token_response_data(access_token)
-
-        if data is not None and data.get('response_type') == 'token':
-            basepath = data.get("redirect_uri")
-            if not basepath:
-                basepath = access_token.client.redirect_uri
-            if len(data.get('state', '')) > 0:
-                response_data['state'] = data.get('state')
-            path = "{0}#{1}".format(basepath, urlencode(response_data))
-            return HttpResponseRedirect(path)
```
Comment (nasthagiri) [inline provider/views.py:552]: Is this code no longer needed inside `access_token_response`?

File: provider/views.py [hunk 4]
```diff
@@ -255,7 +327,36 @@ def error_response(self, request, error, **kwargs):
 
         return self.render_to_response(ctx, **kwargs)
 
+    def get_implicit_response(self, request, client):
```
Comment (nasthagiri) [inline provider/views.py:328]: I'm not too familiar with the details of this flow.  Is there a spec this is following that we can compare against?

File: provider/views.py [hunk 5]
```diff
@@ -255,7 +327,36 @@ def error_response(self, request, error, **kwargs):
...
+            token = self.create_access_token(request, **lookup_kwargs)
+
+        response_data = self.access_token_response_data(token, data['response_type'])
+
+        basepath = data.get("redirect_uri")
+        if not basepath:
+            basepath = token.client.redirect_uri
+        if len(data.get('state', '')) > 0:
+            response_data['state'] = data.get('state')
+        path = "{0}#{1}".format(basepath, urlencode(response_data))
```
Comment (nasthagiri) [inline provider/views.py:353]: Should `#` here be a `?`?

Comment (justinabrahms) [inline provider/views.py:353]: No. This is for the browser-only flow. The `#` is used to prevent data leaking.

You can read more in the oauth spec here: https://tools.ietf.org/html/rfc6749#section-4.2.2

File: provider/views.py (hunk 4 above)
Comment (justinabrahms) [inline provider/views.py:328]: Spec here: https://tools.ietf.org/html/rfc6749#section-4.2

File: provider/oauth2/views.py (hunk 2 above)
Comment (jimabramson) [inline provider/oauth2/views.py]: why exactly?

File: provider/views.py (hunk 3 above)
Comment (jimabramson) [inline provider/views.py:552]: This code was added on the AccessToken view recently, specifically to support implicit flow, which is signified by `response_type='token'` in the authorization request.  Part of this cleanup was to move this logic into `get_implicit_response` because it actually doesn't make sense to have it in the AccessToken view (because 'response type' is not a valid parameter for access token requests).   (short answer - yes, it is no longer needed)

**Comment by jimabramson** [comment]
@clintonb @justinabrahms @nasthagiri coverage improved / satisfied.  i've modified a number of tests to run against both code and implicit flows.  ready for re-review.

File: provider/oauth2/views.py (hunk 2 above)
Comment (clintonb) [inline provider/oauth2/views.py]: In case we ever need a somewhat accurate expiration time. I can imagine debugging an access token issue only to discover hours later that the token wasn't _really_ expired. This looks like one of those things that is no big deal...until it is.

Comment (jimabramson) [inline provider/oauth2/views.py]: ahhh.  i actually thought you were asking for 1d + 1ms, not just 1ms  :-)    will change

Comment (clintonb) [inline provider/oauth2/views.py]: Haha! You have my permission to punch me if I ever make such a silly request and mean it!

File: provider/oauth2/tests.py [hunk 6]
```diff
@@ -113,22 +115,19 @@ def test_authorization_requires_response_type(self):
...
 
-        self.assertEqual(400, response.status_code)
-        self.assertTrue(escape(u"'unsupported' is not a supported response type.") in response.content)
-
-        response = self.client.get(self.auth_url(), data=self.get_auth_params())
-        response = self.client.get(self.auth_url2())
-        self.assertEqual(200, response.status_code, response.content)
+        if response_type == 'unsupported':
+            self.assertEqual(400, response.status_code)
+            self.assertTrue(escape(u"'unsupported' is not a supported response type.") in response.content)
```
Comment (clintonb) [inline provider/oauth2/tests.py:150]: Why are you escaping HTML here?

Comment (jimabramson) [inline provider/oauth2/tests.py:150]: i just moved that while refactoring an existing test.   i don't know why it's needed, but it is used elsewhere when checking error message output (in several tests i have not touched)

**Comment by clintonb** [comment]
:+1:

**Comment by jimabramson** [comment]
rebased and changed -1d expiration to -1ms

**Comment by jimabramson** [comment]
@nasthagiri any outstanding concerns on this PR?

File: provider/views.py [hunk 7]
```diff
@@ -255,7 +325,36 @@ def error_response(self, request, error, **kwargs):
...
+        if constants.SINGLE_ACCESS_TOKEN:
+            token = self.get_access_token(request, **lookup_kwargs)
+        else:
+            token = self.create_access_token(request, **lookup_kwargs)
+
+        response_data = self.access_token_response_data(token, data['response_type'])
+
+        basepath = data.get("redirect_uri")
+        if not basepath:
+            basepath = token.client.redirect_uri
```
Comment (nasthagiri) [inline provider/views.py:350]: According to the spec, the `redirect_ui` should be validated:

```
The authorization server validates the request to ensure that all
   required parameters are present and valid.  The authorization server
   MUST verify that the redirection URI to which it will redirect the
   access token matches a redirection URI registered by the client as
   described in Section 3.1.2.
```

I recommend we require that the redirect_uri in the request (if present) match the previously registered one that is associated with the `Client` in the database.

File: provider/views.py [hunk 8]
```diff
@@ -270,50 +369,31 @@ def handle(self, request, post_data=None):
...
+                "error_description": "The user denied the scope request.",
+                "state": data.get('state'),
+            }
+
+            url = "{0}?{1}".format(
+                client.redirect_uri, urllib.urlencode(response_data))
+
+            return HttpResponseRedirect(url)
+
+        if 'token' in data['response_type']:
```
Comment (nasthagiri) [inline provider/views.py:392]: Why was this changed from an equality check to a membership check?

File: provider/views.py [hunk 9]
```diff
@@ -255,7 +325,36 @@ def error_response(self, request, error, **kwargs):
...
+        """
+        Assumes validated and authorized request with a response_type that is
+        associated with an implicit flow.
+        """
+        data = self.get_data(request)
+
+        lookup_kwargs = {
+            "user": request.user,
+            "client": client,
+            "scope": scope.to_int(*data.get('scope', '').split())
```
Comment (nasthagiri) [inline provider/views.py]: Per [Section 3.3](https://tools.ietf.org/html/rfc6749#section-3.3): 

```
If the client omits the scope parameter when requesting
   authorization, the authorization server MUST either process the
   request using a pre-defined default value or fail the request
   indicating an invalid scope.  The authorization server SHOULD
   document its scope requirements and default value (if defined).
```

Is this documented somewhere?

File: provider/views.py (hunk 7 above)
Comment (nasthagiri) [inline provider/views.py:350]: nvm.  I just saw your response to the earlier comment regarding this being checked in the `AuthorizationRequestForm`'s clean method.

**Comment by nasthagiri** [comment]
Question: Is there a unit test specifically for the implicit grant path?

File: provider/oauth2/tests.py [hunk 10]
```diff
@@ -64,6 +65,7 @@ def _login_and_authorize(self, url_func=None):
         self.assertTrue(self.redirect_url() in response['Location'])
 
 
+@ddt.ddt
```
Comment (nasthagiri) [inline provider/oauth2/tests.py:68]: Please add a test case to verify that a `refresh_token` is _not_ returned when the response_type is 'token' (and the implicit flow is used).

**Comment by nasthagiri** [comment]
Completed my pass.  Just a few questions and a request for additional tests on the implicit flow.

**Comment by nasthagiri** [comment]
Completed my pass.  Just a few questions and a request for additional tests on the implicit flow.

File: provider/views.py (hunk 8 above)
Comment (jimabramson) [inline provider/views.py:392]: extensions to oauth2 are allowed to specify multiple response types, per https://tools.ietf.org/html/rfc6749#section-3.1.1 .  This is important for OIDC which allows "token" to occur in combination with other response types, e.g. "id_token token".

I admit I'm gaming the system a little here, as OIDC also allows "id_token" (without "token") as a response type, which will be caught by this check, which is intentional.  Something like "foo_token" will NOT work, because the response type is still validated prior to this point against a list of allowed response types, which in this package includes only "code" and "token".

File: provider/views.py (hunk 7 above)
Comment (nasthagiri) [inline provider/views.py:350]: One other thing regarding the `redirect_url` (perhaps can be addressed in a separate PR), should we enforce that the supplied `redirect_url` is TLS-protected?  Otherwise, the access token will be sent in the clear.  See the recommendation in http://tools.ietf.org/html/rfc6819#section-4.4.2.1.  @jibsheet FYI - somewhat related to the `SESSION_COOKIE_SECURE` issue.

File: provider/views.py (hunk 8 above)
Comment (nasthagiri) [inline provider/views.py:392]: Ah.  Thanks for the explanation.  To be safer, consider adding a `.split(" ")` at the end.  Or, even better, update `clean_response_type` to return a list of response_types instead of the space-separated string.

File: provider/views.py (hunk 9 above)
Comment (jimabramson) [inline provider/views.py]: thanks for asking this question @nasthagiri as it revealed a bug in the default scope under implicit flow (will push up a fix + test shortly).

as far as the default scope - no it is not "documented" other than by reading / understanding the code, and observing the convention `constants.SCOPES[0]`.  I'll add a one-liner making that a bit more evident.  Note that we aren't making specific use of the 'scope' parameter for any edx stuff, AFAIK.

File: provider/oauth2/tests.py (hunk 10 above)
Comment (jimabramson) [inline provider/oauth2/tests.py:68]: will do

**Comment by jimabramson** [comment]
@nasthagiri updated with additional tests as requested, and added a comment about the default scope.

**Comment by jimabramson** [comment]
@nasthagiri sorry, i meant to do that as a fixup commit but got a little trigger-happy, and the diff isn't helping.  here's what's updated:

test implicit flow (happy path)
https://github.com/edx/django-oauth2-provider/pull/22/files#diff-e1caff531e403a438a4b986347fac835R100

test default scope
https://github.com/edx/django-oauth2-provider/pull/22/files#diff-e1caff531e403a438a4b986347fac835R277

demarcate default scope
https://github.com/edx/django-oauth2-provider/pull/22/files#diff-29026f8c3d3ff3260cdb0cbe3524635dR21

**Comment by nasthagiri** [comment]
:+1:

**Comment by rlucioni** [comment]
@jimabramson don't forget to bump the version number for this package! You can do so in https://github.com/edx/django-oauth2-provider/blob/edx/provider/__init__.py.

**Comment by jimabramson** [comment]
nice catch @rlucioni thanks.  bumping to 0.4.0
